test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs

```diff
@@ -447,6 +456,1909 @@ public async Task GetInstalledVersion_WithAssetsFile_ChangingPackageSpec_Returns
+                        var projectRestoreInfo = ProjectJsonTestHelpers.GetDGSpecFromPackageSpecs(packageSpec);
+                        var projectNames = GetTestProjectNames(projectFullPaths[i], $"project{i}");
+                        projectCache.AddProjectRestoreInfo(projectNames, projectRestoreInfo, new List<IAssetsLogMessage>());
+                    }
+
+                    var initialInstalledPackages = await netCorePackageReferenceProjects[numberOfProjects-1].GetInstalledPackagesAsync(CancellationToken.None);
+
+                    var targetProjects = new List<NetCorePackageReferenceProject>()
+                    {
+                        netCorePackageReferenceProjects[numberOfProjects-1] // Top parent project.
+                    };
+
+                    // Act
+                    var results = await nuGetPackageManager.PreviewProjectsInstallPackageAsync(
+                        targetProjects,
+                        packageB_UpgradeVersion,
+                        resolutionContext,
+                        testNuGetProjectContext,
+                        sourceRepositoryProvider.GetRepositories(),
+                        CancellationToken.None);
+                    var actions = results.Select(a => a.Action).ToArray();
+                    await nuGetPackageManager.ExecuteNuGetProjectActionsAsync(
+                        netCorePackageReferenceProjects,
+                        actions,
+                        testNuGetProjectContext,
+                        new SourceCacheContext(),
+                        CancellationToken.None);
+
+                    // Assert
+                    Assert.Equal(initialInstalledPackages.Count(), 2);
+                    var builtIntegratedActions = actions.OfType<BuildIntegratedProjectAction>().ToList();
+                    Assert.Equal(actions.Count(), targetProjects .Count());
+                    Assert.Equal(actions.Count(), builtIntegratedActions.Count());
+                    // Uprade succeed for this top parent project(no parent but with childs).
+                    // Keep existing Upgrade/downgrade of individual project logic and making sure that my change is not breaking it.
+                    Assert.Equal(actions.Count(), builtIntegratedActions.Count());
+                    Assert.True(builtIntegratedActions.All(b => b.RestoreResult.Success));
+                    Assert.True(builtIntegratedActions.All(b => !b.RestoreResult.LogMessages.Any())); // There should be no error or warning.
+                }
+            }
+            finally
+            {
+                foreach (TestDirectory projectDirectory in projectDirectories)
+                {
+                    projectDirectory.Dispose();
+                }
+            }
+        }
+
+        [Fact]
+        public async Task TestPacMan_PreviewBuildIntegratedProjectsActionsAsync_UpgradePackageFor_MiddleParentProject_Success()
```

Comment:
naming: 

All the names of the tests are `PreviewBuildIntegratedProjectsActionsAsync`, but there's actually no such method ;)

Reply:
Just renamed to `TestPackageManager_InstallPackageForAllProjects_Success` etc...